components/script/dom/htmlimageelement.rs

```diff
@@ -1366,13 +1422,20 @@ impl MicrotaskRunnable for ImageElementMicrotask {
             } => {
                 elem.react_to_environment_changes_sync_steps(*generation);
             },
+            &ImageElementMicrotask::DecodeTask {
+                ref elem,
+                ref promise,
+            } => {
+                elem.decode_image_data_sync(promise);
```

Comment:
So, here we need to break this into two parts:

1. Step 2.1, which run immediately in the microtask.
2. Step 2.2, which runs "in parallel". 

For 2, we need to integrate with the extisting async worflow with the image cache that is setup within [`react_to_environment_changes_sync_steps`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/htmlimageelement.rs#L1024) and which continues [here](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/htmlimageelement.rs#L492). 

I think this is best done by simply adding some state to the image element, so that a pending decoding can be noted and later continued. See for inspiration what is done with [`pending_request`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/htmlimageelement.rs#L166). 

This is actually simpler than I thought, because you don't need to do anything special with the image cache: just keep the decode promise, and then inside [`process_image_response_for_environment_change`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/htmlimageelement.rs#L492C8-L492C53), do the steps in 2.2(except the actual decoding, which happens in the image cache).

Reply:
> So, here we need to break this into two parts:
> 
> 1. Step 2.1, which run immediately in the microtask.
> 2. Step 2.2, which runs "in parallel".
> 
> For 2, we need to integrate with the extisting async worflow with the image cache that is setup within [`react_to_environment_changes_sync_steps`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/htmlimageelement.rs#L1024) and which continues [here](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/htmlimageelement.rs#L492).
> 
> I think this is best done by simply adding some state to the image element, so that a pending decoding can be noted and later continued. See for inspiration what is done with [`pending_request`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/htmlimageelement.rs#L166).
> 
> This is actually simpler than I thought, because you don't need to do anything special with the image cache: just keep the decode promise, and then inside [`process_image_response_for_environment_change`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/htmlimageelement.rs#L492C8-L492C53), do the steps in 2.2(except the actual decoding, which happens in the image cache).

@gterzian I moved decode to `process_image_response_for_environment_change`, please review again, thanks!